File: ptern-typescript/src/semantic/resolver.ts
import {
  type Atom,
  type Capture,
  type Definition,
  type Exclusion,
  type Expression,
  type ParsedPtern,
  type RangeItem,
  type Repetition,
  type Sequence,
} from "../parser/ast";
import { type SemanticError } from "./error";

export function resolve(ptern: ParsedPtern): SemanticError[] {
  // 1. Collect definition names; flag duplicates.
  const [defNames, dupDefErrs] = collectDefNames(ptern.definitions);

  // 2. Detect circular definitions.
  const circErrs = findCircularDefinitions(ptern.definitions, defNames);

  // 3. Check for undefined interpolations inside definition bodies.
  const defRefErrs = ptern.definitions.flatMap(def =>
    checkUndefinedRefs(def.body, defNames, []),
  );

  // 4. Collect capture names in the body.
  const bodyCapNames = capturesInExpression(ptern.body);

  // 5. Flag duplicate capture names in the body.
  const dupCapErrs = findDuplicateNames(bodyCapNames, "duplicateCapture");

  // 6. Collect capture names in all definition bodies (for conflict check).
  const defCapNames = ptern.definitions.flatMap(def => capturesInExpression(def.body));

  // 7. Flag any capture name that collides with a definition name.
  const allCapNames = [...bodyCapNames, ...defCapNames];
  const conflictErrs = findCaptureDefConflicts(allCapNames, defNames);

  // 8. Check for undefined interpolations in the body.
  const bodyRefErrs = checkUndefinedRefs(ptern.body, defNames, bodyCapNames);

  // 9. Flag definitions never reachable from the body expression.
  const unusedErrs = findUnusedDefinitions(ptern.definitions, defNames, ptern.body);

  return [
    ...dupDefErrs,
    ...circErrs,
    ...defRefErrs,
    ...dupCapErrs,
    ...conflictErrs,
    ...bodyRefErrs,
    ...unusedErrs,
  ];
}

// ---------------------------------------------------------------------------
// Definition name collection
// ---------------------------------------------------------------------------

function collectDefNames(defs: Definition[]): [string[], SemanticError[]] {
  const names = defs.map(d => d.name);
  const errors = findDuplicateNames(names, "duplicateDefinition");
  return [dedupList(names), errors];
}

// ---------------------------------------------------------------------------
// Circular definition detection
// ---------------------------------------------------------------------------

function findCircularDefinitions(defs: Definition[], defNames: string[]): SemanticError[] {
  const graph = new Map<string, string[]>();
  for (const def of defs) {
    const deps = interpolationsInExpression(def.body).filter(d => defNames.includes(d));
    graph.set(def.name, deps);
  }

  const allCycles = defNames.flatMap(name => dfsCycles(graph, name, []));
  const sortedCycles = allCycles.map(cycle => [...cycle].sort());
  const uniqueCycles = dedupListOfLists(sortedCycles);
  return uniqueCycles.map(names => ({ kind: "circularDefinition", names }) as SemanticError);
}

function dfsCycles(
  graph: Map<string, string[]>,
  node: string,
  path: string[],
): string[][] {
  if (path.includes(node)) return [takeUntilInclusive(path, node)];
  const newPath = [node, ...path];
  const deps = graph.get(node) ?? [];
  return deps.flatMap(dep => dfsCycles(graph, dep, newPath));
}

function takeUntilInclusive(lst: string[], target: string): string[] {
  const result: string[] = [];
  for (const x of lst) {
    result.push(x);
    if (x === target) break;
  }
  return result;
}

// ---------------------------------------------------------------------------
// Unused definition detection
// ---------------------------------------------------------------------------

function findUnusedDefinitions(
  defs: Definition[],
  defNames: string[],
  body: Expression,
): SemanticError[] {
  const graph = new Map<string, string[]>();
  for (const def of defs) {
    const deps = interpolationsInExpression(def.body).filter(d => defNames.includes(d));
    graph.set(def.name, deps);
  }

  const seeds = interpolationsInExpression(body).filter(n => defNames.includes(n));
  const reachable = expandReachable(graph, seeds, []);

  return defNames
    .filter(name => !reachable.includes(name))
    .map(name => ({ kind: "unusedDefinition", name }) as SemanticError);
}

function expandReachable(
  graph: Map<string, string[]>,
  frontier: string[],
  visited: string[],
): string[] {
  if (frontier.length === 0) return visited;
  const [name, ...rest] = frontier as [string, ...string[]];
  if (visited.includes(name)) return expandReachable(graph, rest, visited);
  const deps = graph.get(name) ?? [];
  return expandReachable(graph, [...deps, ...rest], [name, ...visited]);
}

// ---------------------------------------------------------------------------
// Undefined reference checking
// ---------------------------------------------------------------------------

function checkUndefinedRefs(
  expr: Expression,
  defNames: string[],
  capNames: string[],
): SemanticError[] {
  return interpolationsInExpression(expr)
    .filter(name => !defNames.includes(name) && !capNames.includes(name))
    .map(name => ({ kind: "undefinedReference", name }) as SemanticError);
}

// ---------------------------------------------------------------------------
// Capture / definition name conflict checking
// ---------------------------------------------------------------------------

function findCaptureDefConflicts(capNames: string[], defNames: string[]): SemanticError[] {
  return dedupList(capNames)
    .filter(name => defNames.includes(name))
    .map(name => ({ kind: "captureDefinitionConflict", name }) as SemanticError);
}

// ---------------------------------------------------------------------------
// Collecting interpolation names from the AST
// ---------------------------------------------------------------------------

function interpolationsInExpression(expr: Expression): string[] {
  return expr.alternatives.flatMap(interpolationsInSequence);
}

function interpolationsInSequence(seq: Sequence): string[] {
  return seq.items.flatMap(interpolationsInCapture);
}

function interpolationsInCapture(cap: Capture): string[] {
  return interpolationsInRepetition(cap.inner);
}

function interpolationsInRepetition(rep: Repetition): string[] {
  return interpolationsInExclusion(rep.inner);
}

function interpolationsInExclusion(excl: Exclusion): string[] {
  const base = interpolationsInRangeItem(excl.base);
  const rest = excl.excluded !== null ? interpolationsInRangeItem(excl.excluded) : [];
  return [...base, ...rest];
}

function interpolationsInRangeItem(item: RangeItem): string[] {
  if (item.kind === "singleAtom") return interpolationsInAtom(item.atom);
  return [];
}

function interpolationsInAtom(atom: Atom): string[] {
  switch (atom.kind) {
    case "literal":
    case "charClass":
    case "positionAssertion":
      return [];
    case "interpolation":
      return [atom.name];
    case "group":
      return interpolationsInExpression(atom.inner);
  }
}

// ---------------------------------------------------------------------------
// Collecting capture names from the AST
// ---------------------------------------------------------------------------

function capturesInExpression(expr: Expression): string[] {
  return expr.alternatives.flatMap(capturesInSequence);
}

function capturesInSequence(seq: Sequence): string[] {
  return seq.items.flatMap(capturesInCapture);
}

function capturesInCapture(cap: Capture): string[] {
  const own = cap.name !== null ? [cap.name] : [];
  return [...own, ...capturesInRepetition(cap.inner)];
}

function capturesInRepetition(rep: Repetition): string[] {
  return capturesInExclusion(rep.inner);
}

function capturesInExclusion(excl: Exclusion): string[] {
  return capturesInRangeItem(excl.base);
}

function capturesInRangeItem(item: RangeItem): string[] {
  if (item.kind === "singleAtom") return capturesInAtom(item.atom);
  return [];
}

function capturesInAtom(atom: Atom): string[] {
  if (atom.kind === "group") return capturesInExpression(atom.inner);
  return [];
}

// ---------------------------------------------------------------------------
// List utilities
// ---------------------------------------------------------------------------

function findDuplicateNames(
  names: string[],
  kind: "duplicateDefinition" | "duplicateCapture",
): SemanticError[] {
  const seen = new Set<string>();
  const dups = new Set<string>();
  for (const name of names) {
    if (seen.has(name)) dups.add(name);
    else seen.add(name);
  }
  return [...dups].map(name => ({ kind, name }) as SemanticError);
}

function dedupList(lst: string[]): string[] {
  const seen = new Set<string>();
  const result: string[] = [];
  for (const x of lst) {
    if (!seen.has(x)) {
      seen.add(x);
      result.push(x);
    }
  }
  return result;
}

function dedupListOfLists(lsts: string[][]): string[][] {
  const result: string[][] = [];
  for (const lst of lsts) {
    if (!result.some(r => r.length === lst.length && r.every((x, i) => x === lst[i]))) {
      result.push(lst);
    }
  }
  return result;
}

File: ptern-kotlin/src/main/kotlin/io/ptern/semantic/Resolver.kt
package io.ptern.semantic

import io.ptern.parser.ast.*

object Resolver {
    fun resolve(ptern: ParsedPtern): List<SemanticError> {
        val (defNames, dupDefErrs) = collectDefNames(ptern.definitions)
        val circErrs = findCircularDefinitions(ptern.definitions, defNames)
        val defRefErrs = ptern.definitions.flatMap { checkUndefinedRefs(it.body, defNames, emptyList()) }
        val bodyCapNames = capturesInExpression(ptern.body)
        val dupCapErrs = findDuplicateNames(bodyCapNames) { SemanticError.DuplicateCapture(it) }
        val defCapNames = ptern.definitions.flatMap { capturesInExpression(it.body) }
        val allCapNames = bodyCapNames + defCapNames
        val conflictErrs = findCaptureDefConflicts(allCapNames, defNames)
        val bodyRefErrs = checkUndefinedRefs(ptern.body, defNames, bodyCapNames)
        val unusedErrs = findUnusedDefinitions(ptern.definitions, defNames, ptern.body)

        return dupDefErrs + circErrs + defRefErrs + dupCapErrs + conflictErrs + bodyRefErrs + unusedErrs
    }
}

// ---------------------------------------------------------------------------
// Definition name collection
// ---------------------------------------------------------------------------

private fun collectDefNames(defs: List<Definition>): Pair<List<String>, List<SemanticError>> {
    val names = defs.map { it.name }
    val errors = findDuplicateNames(names) { SemanticError.DuplicateDefinition(it) }
    return Pair(dedup(names), errors)
}

// ---------------------------------------------------------------------------
// Circular definition detection
// ---------------------------------------------------------------------------

private fun findCircularDefinitions(defs: List<Definition>, defNames: List<String>): List<SemanticError> {
    val graph = defs.associate { def ->
        def.name to interpolationsInExpression(def.body).filter { it in defNames }
    }
    val allCycles = defNames.flatMap { dfsCycles(graph, it, emptyList()) }
    val sortedCycles = allCycles.map { it.sorted() }
    return dedupLists(sortedCycles).map { SemanticError.CircularDefinition(it) }
}

private fun dfsCycles(graph: Map<String, List<String>>, node: String, path: List<String>): List<List<String>> {
    if (node in path) return listOf(takeUntilInclusive(path, node))
    val newPath = listOf(node) + path
    return (graph[node] ?: emptyList()).flatMap { dfsCycles(graph, it, newPath) }
}

private fun takeUntilInclusive(lst: List<String>, target: String): List<String> {
    val result = mutableListOf<String>()
    for (x in lst) {
        result.add(x)
        if (x == target) break
    }
    return result
}

// ---------------------------------------------------------------------------
// Unused definition detection
// ---------------------------------------------------------------------------

private fun findUnusedDefinitions(
    defs: List<Definition>,
    defNames: List<String>,
    body: Expression,
): List<SemanticError> {
    val graph = defs.associate { def ->
        def.name to interpolationsInExpression(def.body).filter { it in defNames }
    }
    val seeds = interpolationsInExpression(body).filter { it in defNames }
    val reachable = expandReachable(graph, seeds, emptyList())
    return defNames.filter { it !in reachable }.map { SemanticError.UnusedDefinition(it) }
}

private fun expandReachable(
    graph: Map<String, List<String>>,
    frontier: List<String>,
    visited: List<String>,
): List<String> {
    if (frontier.isEmpty()) return visited
    val name = frontier[0]
    val rest = frontier.drop(1)
    if (name in visited) return expandReachable(graph, rest, visited)
    val deps = graph[name] ?: emptyList()
    return expandReachable(graph, deps + rest, listOf(name) + visited)
}

// ---------------------------------------------------------------------------
// Undefined reference checking
// ---------------------------------------------------------------------------

private fun checkUndefinedRefs(
    expr: Expression,
    defNames: List<String>,
    capNames: List<String>,
): List<SemanticError> =
    interpolationsInExpression(expr)
        .filter { it !in defNames && it !in capNames }
        .map { SemanticError.UndefinedReference(it) }

// ---------------------------------------------------------------------------
// Capture / definition conflict checking
// ---------------------------------------------------------------------------

private fun findCaptureDefConflicts(capNames: List<String>, defNames: List<String>): List<SemanticError> =
    dedup(capNames).filter { it in defNames }.map { SemanticError.CaptureDefinitionConflict(it) }

// ---------------------------------------------------------------------------
// Collecting interpolation names from AST
// ---------------------------------------------------------------------------

private fun interpolationsInExpression(expr: Expression): List<String> =
    expr.alternatives.flatMap { interpolationsInSequence(it) }

private fun interpolationsInSequence(seq: Sequence): List<String> =
    seq.items.flatMap { interpolationsInCapture(it) }

private fun interpolationsInCapture(cap: Capture): List<String> =
    interpolationsInRepetition(cap.inner)

private fun interpolationsInRepetition(rep: Repetition): List<String> =
    interpolationsInExclusion(rep.inner)

private fun interpolationsInExclusion(excl: Exclusion): List<String> =
    interpolationsInRangeItem(excl.base) +
        (excl.excluded?.let { interpolationsInRangeItem(it) } ?: emptyList())

private fun interpolationsInRangeItem(item: RangeItem): List<String> =
    if (item is RangeItem.SingleAtom) interpolationsInAtom(item.atom) else emptyList()

private fun interpolationsInAtom(atom: Atom): List<String> = when (atom) {
    is Atom.Interpolation -> listOf(atom.name)
    is Atom.Group -> interpolationsInExpression(atom.inner)
    else -> emptyList()
}

// ---------------------------------------------------------------------------
// Collecting capture names from AST
// ---------------------------------------------------------------------------

private fun capturesInExpression(expr: Expression): List<String> =
    expr.alternatives.flatMap { capturesInSequence(it) }

private fun capturesInSequence(seq: Sequence): List<String> =
    seq.items.flatMap { capturesInCapture(it) }

private fun capturesInCapture(cap: Capture): List<String> =
    (if (cap.name != null) listOf(cap.name) else emptyList()) + capturesInRepetition(cap.inner)

private fun capturesInRepetition(rep: Repetition): List<String> =
    capturesInExclusion(rep.inner)

private fun capturesInExclusion(excl: Exclusion): List<String> =
    capturesInRangeItem(excl.base)

private fun capturesInRangeItem(item: RangeItem): List<String> =
    if (item is RangeItem.SingleAtom) capturesInAtom(item.atom) else emptyList()

private fun capturesInAtom(atom: Atom): List<String> =
    if (atom is Atom.Group) capturesInExpression(atom.inner) else emptyList()

// ---------------------------------------------------------------------------
// List utilities
// ---------------------------------------------------------------------------

private fun findDuplicateNames(names: List<String>, make: (String) -> SemanticError): List<SemanticError> {
    val seen = mutableSetOf<String>()
    val dups = mutableSetOf<String>()
    for (name in names) {
        if (!seen.add(name)) dups.add(name)
    }
    return dups.map(make)
}

private fun dedup(lst: List<String>): List<String> {
    val seen = mutableSetOf<String>()
    return lst.filter { seen.add(it) }
}

private fun dedupLists(lsts: List<List<String>>): List<List<String>> {
    val result = mutableListOf<List<String>>()
    for (lst in lsts) {
        if (result.none { it.size == lst.size && it.zip(lst).all { (a, b) -> a == b } }) {
            result.add(lst)
        }
    }
    return result
}
